packages/@react-aria/interactions/test/useInteractOutside.test.js

```diff
@@ -205,4 +207,55 @@ describe('useInteractOutside', function () {
       expect(onInteractOutside).not.toHaveBeenCalled();
     });
   });
+
+  describe('shadow dom support', function () {
+    function App({onInteractOutside}) {
+      const ref = useRef(null);
+      useInteractOutside({
+        ref,
+        onInteractOutside
+      });
+
+      return (<div>
+        <div id="outside">I am outside</div>
+        <div id="inside" ref={ref}>I am inside</div>
+      </div>);
+    }
+
+    function Example({onInteractOutside}) {
+      React.useEffect(() => {
+        const host = document.getElementById('shadowHost');
+        if (!!host.shadowRoot && !!(host.shadowRoot).getElementById('mountDiv')) {
+          return;
+        }
+
+        host.attachShadow({mode: 'open'});
+        const mountDiv = document.createElement('div');
+        host.shadowRoot.appendChild(mountDiv);
+
+        ReactDOM.render(
+          <App onInteractOutside={onInteractOutside} />,
+          mountDiv
+        );
+      }, [onInteractOutside]);
+
+      return (<div id="shadowHost" />);
+    }
+
+    it('does not trigger if clicking inside and triggers if clicking outside', function () {
+      const onInteractOutside = jest.fn();
+      render(<Example onInteractOutside={onInteractOutside} />);
+
+      const shadowDom = document.getElementById('shadowHost').shadowRoot;
+      const inside = shadowDom.querySelector('#inside');
+      fireEvent.mouseDown(inside);
+      fireEvent.mouseUp(inside);
+      expect(onInteractOutside).not.toHaveBeenCalled();
+
+      const outside = shadowDom.querySelector('#outside');
+      fireEvent.mouseDown(outside);
+      fireEvent.mouseUp(outside);
+      expect(onInteractOutside).toHaveBeenCalledTimes(1);
```

Comment:
So this is an interesting test, I didn't realize initially that there were two react trees, one inside the shadow dom which we're interested in and the other is outside. In addition, inside/outside on the target divs confused me with the names, because i originally thought it was trying to reference inside and outside the shadow dom. I think as a result, to make this more manageable to those with less shadow dom experience, we should rename those to something a little more real life, maybe `#popover` and `#outside-popover`? I'm open to suggestions.

I think it might be good to add a button outside as a sibling to `Example` and make sure that calls the useInteractOutside as well. What do you think?

Reply:
@snowystinger thanks for your suggestions. I've improved the implementation by just using `ref.current` and rewrote the tests to aid readability. I've also added one to check that clicking a button outside the shadow root triggers.